### PR TITLE
Export to OME-Zarr

### PR DESCRIPTION
**Pull request by btbest**
Support exporting to OME-Zarr

This PR on its own only enables writing to OME-Zarr through headless, since the export GUI is in volumina (https://github.com/ilastik/volumina/pull/316)

* Local file system only, writing to web stores tbd
* Output image is single-scale. Scaling turns out to be quite complex to do properly and blockwise / RAM-friendly / web-friendly, so leaving it for another PR. Hopefully if someone wants to have their output scaled before I get it done, they can use other tools that are out there. Though I'm not sure how good the converters/scalers out there actually are at handling OME-Zarr as input. Once I do implement it, they will be able to load any single-scale exports from this version, and "convert" them to multiscale by exporting them in the to-be-added format option.
* If the input was already a multiscale dataset (Precomputed or OME-Zarr), the output will match the scale key to the input, and the output's scaling factor metadata will be recorded relative to the raw data size of the input
* If the input was OME-Zarr, the output will carry over scaling and translation metadata both for the entire pyramid and for the exported scale(s). In regards to scaling metadata of individual scales, this is specifically the "t" scaling factor, since the factors for xyz depend on the actual written scale (or scaling relative to raw data, see above), and metadata for "c" should not be carried over because export channels != input channels. Scaling metadata on the pyramid level is supposed to represent pixel resolutions relevant for all scales, so only "t" should ever be anything other than 1.0. But here we carry over all numbers to the output, because who knows what they might mean, and there are no meaningful numbers from ilastik's side to write there instead.
* Refuses to write to existing paths, enable appending and merging metadata tbd

## Checklist

- [x] Format code and imports.
- [x] Add docstrings and comments.
- [x] Add tests.
- [x] Update [user documentation](https://github.com/ilastik/ilastik.github.io). -- https://github.com/ilastik/ilastik.github.io/pull/281
- [x] Reference relevant issues and other pull requests.
- [x] Rebase commits into a logical sequence.

### CODE REVIEW

**Comment by btbest** [comment]
Back in draft

* [ ] Add e2e test
* [ ] Minor refactor in _compute_and_write_scales

**Comment by btbest** [comment]
CI broken due to protobuf update (see https://github.com/ilastik/ilastik/pull/2910), full test run was green on my system

**Comment by btbest** [comment]
Back in draft 😭 When writing to existing stores, if there is already an array at the exact target path, it is currently overwritten (which I think is what we want).

However, when writing a new array to an existing store, to an internal path that did not yet exist, the new array will be written, but the old one will still stay. However, the OME-Zarr metadata at the store root will be overwritten and only describe the new array.

Merging the metadata is somewhat complex (because then the writer has to start exploring the entire store), so in the interest of finally getting this PR done, I will change this to simply refuse to write to existing stores.

**Comment by btbest** [comment]
I think that's actually it 😅 

At this point OME-Zarr feels pretty good to use in ilastik I gotta say.

**Comment by k-dominik** [comment]
Heya, still looking into this, but while trying it out I ran into a problem writing single scale ome-zarr from precomputed chunks

```
ERROR 2024-10-02 14:14:13,173 log_exception 50588 14496002048 Traceback (most recent call last):
  File "..../ilastik/ilastik/applets/dataExport/dataExportGui.py", line 417, in exportSync
    opLaneView.run_export()
  File "..../ilastik/ilastik/applets/dataExport/opDataExport.py", line 248, in run_export
    self._opFormattedExport.run_export()
  File "..../ilastik/lazyflow/operators/ioOperators/opFormattedDataExport.py", line 276, in run_export
    self._opExportSlot.run_export()
  File "..../ilastik/lazyflow/operators/ioOperators/opExportSlot.py", line 273, in run_export
    export_func()
  File "..../ilastik/lazyflow/operators/ioOperators/opExportSlot.py", line 422, in _export_ome_zarr
    write_ome_zarr(self.ExportPath.value, self.Input, self.progressSignal, compute_downscales)
  File "..../ilastik/lazyflow/utility/io_util/write_ome_zarr.py", line 426, in write_ome_zarr
    export_scalings, scalings_relative_to_raw_input = _match_or_create_scalings(
  File "..../ilastik/lazyflow/utility/io_util/write_ome_zarr.py", line 177, in _match_or_create_scalings
    filtered_input_scales = _get_input_multiscales_matching_export(input_scales, export_shape, compute_downscales)
  File "..../ilastik/lazyflow/utility/io_util/write_ome_zarr.py", line 133, in _get_input_multiscales_matching_export
    assert len(matching_scales) > 0, "Should be impossible, input must be one of the scales"
AssertionError: Should be impossible, input must be one of the scales
```

(data conversion workflow, random precomputed image: precomputed://https://open-neurodata.s3.amazonaws.com/collman/collman15v2/EM25K export...)

**Comment by btbest** [comment]
Fix is up; will add a test and rebase on main

**Review by k-dominik — COMMENTED** [review]
Heya @btbest,

I think this PR is in a good state! Code was good to follow and I didn't find anything deep. Also great that you took care to write ome/ngff-zarr such that other applications will actually open it. I for one wasn't aware that `fill_value` is a must there...

In the description of your PR you state that

> If the input was OME-Zarr, the output will carry over scaling and translation metadata both for the entire pyramid and for the exported scale(s).

Is this still up to date? I tried and did not find any metadata for other scales.

Side note, it took me a while to find a zarr that I could export. I tried https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0054A/5025553.zarr and while it was showing just fine, export would time out. I also tried https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0062A/6001240.zarr . Maybe it's a current problem with idr. The latter couldn't even be loaded at all.

I'm not so sure about many of the partially not implemented downscaling methods, like `_apply_scaling_method`, `_round_like_scaling_method`, or all code paths in other functions that would error out now. Sure we have dead code in the repo already, I'm just not sure whether this is a good idea. Also it would reduce the size of this PR and the future PR with downscaling/rescaling would be more concise.

Appreciated the clear and well written docstrings that are very clear about ambiguous terms like resolution and scales - future devs will be thankful :).

**Comment by btbest** [comment]
> In the description of your PR you state that
> 
> > If the input was OME-Zarr, the output will carry over scaling and translation metadata both for the entire pyramid and for the exported scale(s).
> 
> Is this still up to date? I tried and did not find any metadata for other scales.

Just an unfortunate phrasing; this was supposed to say that if the export was able to write more than one scale, it would have carried over metadata from each input scale matching an export scale...

> Side note, it took me a while to find a zarr that I could export. I tried https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0054A/5025553.zarr and while it was showing just fine, export would time out. I also tried https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0062A/6001240.zarr . Maybe it's a current problem with idr. The latter couldn't even be loaded at all.

Seems it was temporary on IDR-side? I'm able to export both now (though it does take a while, and I'm not sure the progress bar feedback is working correctly since it stayed on 0% until it was finished)

> I'm not so sure about many of the partially not implemented downscaling methods, like `_apply_scaling_method`, `_round_like_scaling_method`, or all code paths in other functions that would error out now. Sure we have dead code in the repo already, I'm just not sure whether this is a good idea. Also it would reduce the size of this PR and the future PR with downscaling/rescaling would be more concise.

I suppose I wanted to keep the parts of the implementation that I already considered working - specifically, the logic for deciding if and how to scale. These methods were left as the implementation points for the remaining bits of logic. But true, better to have an internally consistent state of code, so I will removed all of it. Can just revert that commit when I get to the implementation.

Further points on the respective comments as I update :)

File: tests/test_lazyflow/test_operators/test_ioOperators/testOpExportSlot.py [hunk 1]
```diff
@@ -78,6 +81,198 @@ def testBasic_Hdf5(self):
...
+        graph = Graph()
+        opPiper = OpArrayPiper(graph=graph)
+        opPiper.Input.setValue(data)
+
+        opExport = OpExportSlot(graph=graph)
+        opExport.Input.connect(opPiper.Output)
+        opExport.OutputFormat.setValue("single-scale OME-Zarr")
+        opExport.OutputFilenameFormat.setValue(self._tmpdir + "/test_export_x{x_start}-{x_stop}_y{y_start}-{y_stop}")
+        opExport.OutputInternalPath.setValue("")  # Overwrite the slot's default "exported_data"
+        opExport.CoordinateOffset.setValue((10, 20))
```
[the remark this answers is not in the document]
Comment (btbest) [reply tests/test_lazyflow/test_operators/test_ioOperators/testOpExportSlot.py:97]: Err... nice find, there were quite a few parts of the metadata carry-over that would break if the user chose to crop the export...

File: lazyflow/utility/io_util/RESTfulPrecomputedChunkedVolume.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ class RESTfulPrecomputedChunkedVolume(MultiscaleStore):
                     "type": "object",
                     "properties": {
                         "key": {"type": "string"},
-                        "resolution": {"type": "array", "items": {"type": "number"}},
+                        "size": {"type": "array", "items": {"type": "number"}},
```
[the remark this answers is not in the document]
Comment (btbest) [reply lazyflow/utility/io_util/RESTfulPrecomputedChunkedVolume.py:71]: Yes, and I had to stop myself. Precomputed-to-OME-Zarr conversion is not really in scope here

File: lazyflow/utility/io_util/OMEZarrStore.py [hunk 3]
```diff
@@ -41,6 +42,60 @@
...
+OMEZarrCoordinateTransformation = Dict[str, Union[str, List[float]]]
+
+
+def _remove_transforms_with_path(
+    coordinate_transformations: Optional[List[OMEZarrCoordinateTransformation]],
+) -> Optional[List[OMEZarrCoordinateTransformation]]:
+    """
+    coordinateTransformations may provide a path to transformation data in binary format, instead
+    of specifying floats for each axis. We don't know of any commonly used tool that writes such data,
+    so there is no way to know what to do with this.
```
[the remark this answers is not in the document]
Comment (btbest) [reply lazyflow/utility/io_util/OMEZarrStore.py:58]: Bad docstring; the actual problem is that the spec doesn't actually allow including coordinateTransforms with a `"path"`, the same as it does not allow including an `identity` transformation (despite the coordinateTransformation spec defining it as a valid transformation). Quote [multiscale spec](https://ngff.openmicroscopy.org/latest/index.html#multiscale-md):

> The transformation MUST only be of type translation or scale. They MUST contain exactly one scale transformation that specifies the pixel size in physical units or time duration. If scaling information is not available or applicable for one of the axes, the value MUST express the scaling factor between the current resolution and the first resolution for the given axis, defaulting to 1.0 if there is no downsampling along the axis. It MAY contain exactly one translation that specifies the offset from the origin in physical units. If translation is given it MUST be listed after scale to ensure that it is given in physical coordinates. The length of the scale and translation array MUST be the same as the length of "axes".

Given that the scale transformation MUST be present for each `dataset`, and it MUST specify "the pixel size in physical units or time duration", I think a writer would be pretty misguided to think it can somehow create an OME-Zarr store that anyone else can read if it provides the scaling in some arbitrary binary format.

tl;dr I see 0 value spending any more time than changing the docstring on this

File: lazyflow/utility/io_util/OMEZarrStore.py [hunk 4]
```diff
@@ -72,6 +127,10 @@ def _get_zarr_cache_max_size() -> int:
     return math.floor(caches_max * permissible_fraction_max)
 
 
+def scale_key_from_path(scale_path):
+    return scale_path.split("/")[-1]
```
[the remark this answers is not in the document]
Comment (btbest) [reply lazyflow/utility/io_util/OMEZarrStore.py:131]: In fact if it contained a `\`, we would have to assume it's part of the scale key...

File: tests/test_lazyflow/test_utility/test_io_util/test_write_ome_zarr.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+)
+def test_metadata_integrity(tmp_path, graph, shape, axes):
+    data_array = vigra.VigraArray(shape, axistags=vigra.defaultAxistags(axes))
+    data_array[...] = numpy.indices(shape).sum(0)
+    export_path = tmp_path / "test.zarr"
+    source_op = OpArrayPiper(graph=graph)
+    source_op.Input.setValue(data_array)
+    progress = mock.Mock()
+
+    write_ome_zarr(str(export_path), source_op.Output, progress, compute_downscales=True)
```
[the remark this answers is not in the document]
Comment (btbest) [reply tests/test_lazyflow/test_utility/test_io_util/test_write_ome_zarr.py:35]: Deleted for this PR, but I think I might try to rephrase this test as parametrised assert that can be added to the other tests

**Comment by btbest** [comment]
Err... I don't know what's going on with the conda build on CI but oh well

File: lazyflow/utility/io_util/OMEZarrStore.py [hunk 6]
```diff
@@ -41,6 +42,60 @@
 OME_ZARR_V_0_4_KWARGS = dict(dimension_separator="/", normalize_keys=False)
 OME_ZARR_V_0_1_KWARGS = dict(dimension_separator=".")
 
+# {
+#   "type": "scale" OR "translation",
+#   "scale": List[float] OR "translation": List[float] OR "path": str
+# }
+OMEZarrCoordinateTransformation = Dict[str, Union[str, List[float]]]
```
[the remark this answers is not in the document]
Comment (k-dominik) [reply lazyflow/utility/io_util/OMEZarrStore.py]: Ah, yeah I missed that!

> The transformation MUST only be of type translation or scale.

Thanks :)

**Review by k-dominik — APPROVED** [review]
Thx for making the review easy. Additional changes address relevant comments, and I can also understand well, why some comments would not be addressed :).

I could confirm in MoBiE that cutting out a piece of data will end up in the right location in the original dataset - which I absolutely love as a feature :)
